src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Options/Option2`1.cs

```diff
@@ -0,0 +1,135 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System;
+using System.Collections.Immutable;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Options
+{
+    /// <summary>
+    /// Marker interface for language specific options.
+    /// </summary>
+    internal interface ILanguageSpecificOption : IOptionWithGroup
+    {
+    }
+
+    /// <summary>
+    /// Marker interface for language specific options.
+    /// </summary>
+    internal interface ILanguageSpecificOption<T> : ILanguageSpecificOption
+    {
+    }
+
+    /// <summary>
+    /// An global option. An instance of this class can be used to access an option value from an OptionSet.
+    /// </summary>
+    internal class Option2<T> : ILanguageSpecificOption<T>
+    {
+        /// <summary>
+        /// Feature this option is associated with.
+        /// </summary>
+        public string Feature { get; }
+
+        /// <summary>
+        /// Optional group/sub-feature for this option.
+        /// </summary>
+        internal OptionGroup Group { get; }
+
+        /// <summary>
+        /// The name of the option.
+        /// </summary>
+        public string Name { get; }
+
+        /// <summary>
+        /// The default value of the option.
+        /// </summary>
+        public T DefaultValue { get; }
+
+        /// <summary>
+        /// The type of the option value.
+        /// </summary>
+        public Type Type => typeof(T);
+
+        public ImmutableArray<OptionStorageLocation2> StorageLocations { get; }
+
+        [Obsolete("Use a constructor that specifies an explicit default value.")]
+        public Option2(string feature, string name)
+            : this(feature, name, default!)
+        {
+            // This constructor forwards to the next one; it exists to maintain source-level compatibility with older callers.
+        }
```

Comment:
💡  This constructor should be removed since `Option2<T>` is a new internal type.